File: 2017/test/day_18_test.exs
defmodule AdventOfCode2017.Day18Test do
  use ExUnit.Case

  import AdventOfCode2017.Day18
  import AdventOfCode2017.TestHelper

  # test "part1" do
  #   assert part1(sample_input_1()) == 4
  #   # assert part1(input("test/input/day_17.txt")) == 1173
  # end

  # test "snd" do

  # end

  test "set" do
    modified_registers = execute({"set", "a", 1}, sample_registers_1())
    assert modified_registers["a"] == 1
    # assert execute({"set", "a", 1}, sample_registers_1()) == 4
  end

  test "add" do
    modified_registers = execute({"add", "b", 3}, sample_registers_1())
    assert modified_registers["b"] == 3
  end

  test "mul" do
    registers = Map.put(sample_registers_1(), "c", 4)
    modified_registers = execute({"mul", "c", 2}, registers)
    assert modified_registers["c"] == 8
  end

  test "mod" do
    registers = Map.put(sample_registers_1(), "d", 5)
    modified_registers = execute({"mod", "d", 3}, registers)
    assert modified_registers["d"] == 2
  end

  # test "rcv" do

  # end

  # test "jgz" do

  # end

  # test "part2" do
  #   assert part2(input("test/input/day_17.txt")) == 1930815
  # end

  test "parse" do
    assert parse(sample_input_1()) == sample_instructions_1()
  end

  def sample_input_1() do
    "set a 1\nadd a 2\nmul a a\nmod a 5\nsnd a\nset a 0\nrcv a\njgz a -1\nset a 1\njgz a -2"
  end

  def sample_instructions_1() do
    [
      {"set", "a", 1},
      {"add", "a", 2},
      {"mul", "a", "a"},
      {"mod", "a", 5},
      {"snd", "a"},
      {"set", "a", 0},
      {"rcv", "a"},
      {"jgz", "a", -1},
      {"set", "a", 1},
      {"jgz", "a", -2}
    ]
  end

  def sample_registers_1() do
    Map.new(?a..?z, fn reg -> {<<reg :: utf8>>, 0} end)
  end
end

File: 2018/test/day_18_test.rb
require "test_helper"
require "day_18"

class Day18Test < Minitest::Test
  def setup
    @day_18 = Day18.new
  end

  def test_part1
    assert_equal 1147, @day_18.part1(sample_input_1)
    assert_equal 621205, @day_18.part1(input("day_18.txt"))
  end

  def test_part2
    assert_equal 228490, @day_18.part2(input("day_18.txt"))
  end

  def sample_input_1
    ".#.#...|#.
    .....#|##|
    .|..|...#.
    ..|#.....#
    #.#|||#|#|
    ...#.||...
    .|....|...
    ||...#|.#|
    |.||||..|.
    ...#.|..|."
  end
end
